README.md
[![Build Status](https://travis-ci.org/chris48s/cakephp-geodistance.svg?branch=master)](https://travis-ci.org/chris48s/cakephp-geodistance)
[![Coverage Status](https://coveralls.io/repos/github/chris48s/cakephp-geodistance/badge.svg?branch=master)](https://coveralls.io/github/chris48s/cakephp-geodistance?branch=master)

# CakePHP GeoDistance Plugin
## A CakePHP 3 Behavior for querying geocoded data by distance.

CakePHP-GeoDistance is a CakePHP 3 behavior for querying geocoded data based on
cartographic distance using the spherical cosine law. It is great for 'find my
nearest X' or 'find Y near me' type queries. If your database doesn't already
have latitude/longitude co-ordinates attached to your geographic data, you can
add them using a geocoding plugin. Try
[this one](https://github.com/chris48s/cakephp-geocoder).


## Installation

Install from [packagist](https://packagist.org/packages/chris48s/cakephp-geodistance) using [composer](https://getcomposer.org/).
Add the following to your `composer.json`:

```
"require": {
    "chris48s/cakephp-geodistance": "^2.0.0"
}
```

and run `composer install` or `composer update`, as applicable.

## Supported databases

Only MySQL and Postgres are supported.

## Usage

### Loading the plugin

Add the code `Plugin::load('Chris48s/GeoDistance');` to your `bootstrap.php`.

### Using the Behavior

Add the behavior in your table class.

```php
<?php
namespace App\Model\Table;

use Cake\ORM\Table;

class MyTable extends Table
{

    public function initialize(array $config)
    {
        parent::initialize($config);
        $this->addBehavior('Chris48s/GeoDistance.GeoDistance');
    }
}
```

#### Configuration

By default, the behavior assumes your table contains columns called `latitude`
and `longitude`, and you want to perform queries in miles. These can be changed
though. Simply pass an array of options when attaching the behavior:

```php
<?php
namespace App\Model\Table;

use Cake\ORM\Table;

class MyTable extends Table
{

    public function initialize(array $config)
    {
        parent::initialize($config);
        $this->addBehavior('Chris48s/GeoDistance.GeoDistance', [
            'latitudeColumn' => 'lat',
            'longitudeColumn' => 'lng',
            'units' => 'km'
        ]);
    }
}
```

Supported units are miles or kilometres.

#### Querying data

Having added the behavior to a table class, you now have access to the query
method `find('bydistance')`, which you can use to find database records within
`radius` of a given point:

```php
<?php

use Cake\ORM\TableRegistry;

$myTable = TableRegistry::get('MyTable');

$options = [
    'latitude' => 51.3984830139,
    'longitude' => -0.236298886484,
    'radius' => 10
];
$query = $myTable
    ->find('bydistance', $options)
    ->select(['address', 'lat', 'lng']);
```

`latitude`, `longitude` and `radius` are required parameters. If required
parameters are missing or invalid, an exception of class `GeoDistanceInvalidArgumentException`
will be thrown.

You can also pass additional conditions or parameters to the query and
override the default for 'units', for example:

```php
<?php

use Cake\ORM\TableRegistry;

$myTable = TableRegistry::get('MyTable');

$options = [
    'latitude' => 51.3984830139,
    'longitude' => -0.236298886484,
    'radius' => 10,
    'units' => 'kilometres',
    'conditions' => [ 'active' => 1 ]
];
$query = $myTable
    ->find('bydistance', $options)
    ->select(['address', 'lat', 'lng']);
```

The method `find('bydistance')` returns a CakePHP query object, so you can chain
additional methods on to this (e.g: `->order()`, `->limit()`, etc).

## Reporting Issues

If you have any issues with this plugin then please feel free to create a new
[Issue](https://github.com/chris48s/cakephp-geodistance/issues) on the
[GitHub repository](https://github.com/chris48s/cakephp-geodistance).
This plugin is licensed under the MIT License.


CHANGELOG.md
# Changelog

## :package: [2.0.0](https://packagist.org/packages/chris48s/cakephp-geodistance#2.0.0) - 2018-05-06

* Remove deprecated CakePHP calls
* Require CakePHP >=3.4

## :package: [1.1.0](https://packagist.org/packages/chris48s/cakephp-geodistance#1.1.0) - 2017-08-30

* Bugfix: ensure chaining `->select()` works correctly
* Require PHP >=5.6

## :package: [1.0.0](https://packagist.org/packages/chris48s/cakephp-geodistance#1.0.0) - 2017-01-09

First Release


tests/bootstrap.php
<?php
// @codingStandardsIgnoreFile

$findRoot = function () {
    $root = dirname(__DIR__);
    if (is_dir($root . '/vendor/cakephp/cakephp')) {
        return $root;
    }

    $root = dirname(dirname(__DIR__));
    if (is_dir($root . '/vendor/cakephp/cakephp')) {
        return $root;
    }

    $root = dirname(dirname(dirname(__DIR__)));
    if (is_dir($root . '/vendor/cakephp/cakephp')) {
        return $root;
    }
};

if (!defined('DS')) {
    define('DS', DIRECTORY_SEPARATOR);
}
define('ROOT', $findRoot());
define('APP_DIR', 'App');
define('WEBROOT_DIR', 'webroot');
define('APP', ROOT . '/tests/App/');
define('CONFIG', ROOT . '/tests/config/');
define('WWW_ROOT', ROOT . DS . WEBROOT_DIR . DS);
define('TESTS', ROOT . DS . 'tests' . DS);
define('TMP', ROOT . DS . 'tmp' . DS);
define('LOGS', TMP . 'logs' . DS);
define('CACHE', TMP . 'cache' . DS);
define('CAKE_CORE_INCLUDE_PATH', ROOT . '/vendor/cakephp/cakephp');
define('CORE_PATH', CAKE_CORE_INCLUDE_PATH . DS);
define('CAKE', CORE_PATH . 'src' . DS);

require ROOT . '/vendor/autoload.php';
require CORE_PATH . 'config/bootstrap.php';

Cake\Core\Configure::write('App', ['namespace' => 'App']);
Cake\Core\Configure::write('debug', true);

$TMP = new \Cake\Filesystem\Folder(TMP);
$TMP->create(TMP . 'cache/models', 0777);
$TMP->create(TMP . 'cache/persistent', 0777);
$TMP->create(TMP . 'cache/views', 0777);

$cache = [
    'default' => [
        'engine' => 'File'
    ],
    '_cake_core_' => [
        'className' => 'File',
        'prefix' => 'version_myapp_cake_core_',
        'path' => CACHE . 'persistent/',
        'serialize' => true,
        'duration' => '+10 seconds'
    ],
    '_cake_model_' => [
        'className' => 'File',
        'prefix' => 'version_my_app_cake_model_',
        'path' => CACHE . 'models/',
        'serialize' => 'File',
        'duration' => '+10 seconds'
    ]
];

Cake\Cache\Cache::setConfig($cache);
Cake\Core\Configure::write('Session', [
    'defaults' => 'php'
]);

Cake\Core\Plugin::load('Chris48s/GeoDistance', ['path' => ROOT . DS, 'autoload' => true]);

// Ensure default test connection is defined
if (!getenv('db_dsn')) {
    putenv('db_dsn=sqlite:///:memory:');
}

Cake\Datasource\ConnectionManager::setConfig('test', [
    'url' => getenv('db_dsn'),
    'timezone' => 'UTC'
]);


tests/Fixture/FooFixture.php
<?php

namespace Chris48s\GeoDistance\Test\Fixture;

use Cake\TestSuite\Fixture\TestFixture;

class FooFixture extends TestFixture
{
    public $table = 'foo';

    public $fields = [
        'id' => ['type' => 'integer'],
        'name' => ['type' => 'string', 'null' => true],
        'lat' => ['type' => 'float', 'null' => false],
        'lng' => ['type' => 'float', 'null' => false],
        'active' => ['type' => 'boolean', 'null' => false],
        '_constraints' => ['primary' => ['type' => 'primary', 'columns' => ['id']]]
    ];

    public $records = [
        ['id' => 1, 'name' => 'south pole', 'lat' => -90, 'lng' => 0, 'active' => false ],
        ['id' => 2, 'name' => 'north pole', 'lat' => 90, 'lng' => 0, 'active' => false ],
        ['id' => 3, 'name' => 'same 1', 'lat' => 0, 'lng' => -180, 'active' => false ],
        ['id' => 4, 'name' => 'same 2', 'lat' => 0, 'lng' => 180, 'active' => false ],
        ['id' => 5, 'name' => 'birminham centre', 'lat' => 52.47980068128972, 'lng' => -1.8967723846435545, 'active' => true ],
        ['id' => 6, 'name' => 'birmingham close 1', 'lat' => 52.4858640956247, 'lng' => -1.8966865539550781, 'active' => true ],
        ['id' => 7, 'name' => 'birmingham close 2', 'lat' => 52.47985295567416, 'lng' => -1.904325485229492, 'active' => true ],
        ['id' => 8, 'name' => 'birmingham close 3', 'lat' => 52.47718688287627, 'lng' => -1.8944549560546875, 'active' => false ],
        ['id' => 9, 'name' => 'birmingham far', 'lat' => 52.50514646853436, 'lng' => -1.8513679504394531, 'active' => true ],
        ['id' => 10, 'name' => 'wrong birmingham', 'lat' => 33.519644153199245, 'lng' => -86.8033218383789, 'active' => true ],
    ];
}


tests/TestCase/Model/Behavior/GeoDistanceBehaviorTest.php
<?php

namespace Chris48s\GeoDistance\Test\TestCase\Model\Behavior;

use Cake\Datasource\ConnectionManager;
use Cake\ORM\TableRegistry;
use Cake\TestSuite\TestCase;
use Chris48s\GeoDistance\Exception\GeoDistanceFatalException;
use Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException;
use Chris48s\GeoDistance\Model\Behavior\GeoDistanceBehavior;

class GeoDistanceBehaviorTest extends TestCase
{
    public $fixtures = [
        'plugin.Chris48s\GeoDistance.foo'
    ];

    public function tearDown()
    {
        parent::tearDown();
        TableRegistry::clear();
    }

    // set up a table to use for testing
    private function getValidTable()
    {
        $table = TableRegistry::get('Foo');
        $table->addBehavior('Chris48s/GeoDistance.GeoDistance', [
            'latitudeColumn' => 'lat',
            'longitudeColumn' => 'lng'
        ]);

        return $table;
    }

    /* our fixture table does not contain columns called 'latitude' or 'longitude'
       so we expect a GeoDistanceFatalException to be thrown */
    public function testSetupInvalidColumns()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceFatalException');

        $table = TableRegistry::get('Foo');
        $table->addBehavior('Chris48s/GeoDistance.GeoDistance');
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'latitude' is not specified
    public function testInvalidQueryParamsNoLat()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'longitude' => 0,
            'radius' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'longitude' is not specified
    public function testInvalidQueryParamsNoLng()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'radius' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'radius' is not specified
    public function testInvalidQueryParamsNoRadius()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'latitude' is outside valid range
    public function testInvalidQueryParamsBadLat1()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 91,
            'longitude' => 0,
            'radius' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'latitude' is not numeric
    public function testInvalidQueryParamsBadLat2()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 'foo',
            'longitude' => 0,
            'radius' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'longitude' is outside valid range
    public function testInvalidQueryParamsBadLng1()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 181,
            'radius' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'longitude' is not numeric
    public function testInvalidQueryParamsBadLng2()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 'foo',
            'radius' => 0
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // ensure GeoDistanceInvalidArgumentException is thrown if 'radius' is not numeric
    public function testInvalidQueryParamsBadRadius()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 0,
            'radius' => 'foo'
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    /* ensure GeoDistanceInvalidArgumentException is thrown if
       'units' is not 'miles', 'mi', 'kilometres' or 'km' */
    public function testInvalidQueryParamsBadUnits()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException');

        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 0,
            'radius' => 0,
            'units' => 'foo'
        ];
        $query = $table->find('bydistance', $options);
        $query->toArray();
    }

    // check chaining ->select() works correctly
    public function testValidQueryWithSelect()
    {
        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 180,
            'radius' => 0,
            'units' => 'km'
        ];
        $query = $table
            ->find('bydistance', $options)
            ->select(['name', 'lat', 'lng']);
        $result = $query->toArray();
        $this->assertArrayHasKey('name', $result[0]);
        $this->assertArrayHasKey('distance', $result[0]);
        $this->assertArrayNotHasKey('id', $result[0]);
    }

    // check (0, 180) and (0, -180) are considered the same point
    public function testWrapAround()
    {
        $table = $this->getValidTable();
        $options = [
            'latitude' => 0,
            'longitude' => 180,
            'radius' => 0,
            'units' => 'km'
        ];
        $query = $table
            ->find('bydistance', $options)
            ->select(['name', 'lat', 'lng']);
        $result = $query->toArray();
        $this->assertEquals(2, count($result));
    }

    // ensure correct points are returned with a small radius
    public function testSmallDistance()
    {
        $table = $this->getValidTable();
        $options = [
            'latitude' => 52.47980068128972,
            'longitude' => -1.8967723846435545,
            'radius' => 0.9,
            'units' => 'km'
        ];
        $query = $table
            ->find('bydistance', $options)
            ->select(['name', 'lat', 'lng']);
        $result = $query->toArray();
        $this->assertEquals(4, count($result));
        foreach ($result as $row) {
            $this->assertTrue(in_array(
                $row['name'],
                ['birminham centre', 'birmingham close 1', 'birmingham close 2', 'birmingham close 3']
            ));
        }
    }

    // ensure extra conditions are also applied
    public function testExtraConditions()
    {
        $table = $this->getValidTable();
        $options = [
            'latitude' => 52.47980068128972,
            'longitude' => -1.8967723846435545,
            'radius' => 1,
            //'units' => 'miles', is implicit if not specified
            'conditions' => [ 'active' => 1 ]
        ];
        $query = $table
            ->find('bydistance', $options)
            ->select(['name', 'lat', 'lng']);
        $result = $query->toArray();
        $this->assertEquals(3, count($result));
        foreach ($result as $row) {
            $this->assertTrue(in_array(
                $row['name'],
                ['birminham centre', 'birmingham close 1', 'birmingham close 2']
            ));
        }
    }

    /* every point on the earth's surface should be a max of
       round((pi() * (2 * 6371)) / 2, 3)km away from every other point
       based on 6371km as an approximation of the earth's mean radius */
    public function testLargeDistance()
    {
        $table = $this->getValidTable();
        $options = [
            'latitude' => 90,
            'longitude' => 0,
            'radius' => round((pi() * (2 * 6371)) / 2, 3),
            'units' => 'km'
        ];
        $query = $table
            ->find('bydistance', $options)
            ->select(['name', 'lat', 'lng']);
        $result = $query->toArray();
        $this->assertEquals(10, count($result));
        foreach ($result as $row) {
            if ($row['name'] == 'south pole') {
                $this->assertEquals(round((pi() * (2 * 6371)) / 2, 3), $row['distance']);
            }
        }
    }

    /* Only MySQL and Postgres are supported
       Pass in a connection to another DB engine
       and ensure correct exception is thrown */
    public function testInvalidDB()
    {
        $this->setExpectedException('Chris48s\GeoDistance\Exception\GeoDistanceFatalException');

        //set up a SQLite DB connection - SQLite is not supported
        ConnectionManager::setConfig('invalid', [
            'url' => 'sqlite:///:memory:',
            'timezone' => 'UTC'
        ]);
        $conn = ConnectionManager::get('invalid');

        //create a table in SQLite
        $conn->query("CREATE TABLE `Foo` (
            `id` int(11) NOT NULL,
            `lat` float,
            `lng` float,
            PRIMARY KEY (`id`)
        );");
        $table = TableRegistry::get('Foo', ['connection' => $conn]);
        $table->addBehavior('Chris48s/GeoDistance.GeoDistance', [
            'latitudeColumn' => 'lat',
            'longitudeColumn' => 'lng'
        ]);

        //tidy up
        ConnectionManager::dropAlias('invalid');
    }
}


src/Model/Behavior/GeoDistanceBehavior.php
<?php

namespace Chris48s\GeoDistance\Model\Behavior;

use Cake\Database\Driver\Mysql;
use Cake\Database\Driver\Postgres;
use Cake\ORM\Behavior;
use Cake\ORM\Query;
use Chris48s\GeoDistance\Exception\GeoDistanceFatalException;
use Chris48s\GeoDistance\Exception\GeoDistanceInvalidArgumentException;

class GeoDistanceBehavior extends Behavior
{
    protected $_defaultConfig = [
        'latitudeColumn' => 'latitude',
        'longitudeColumn' => 'longitude',
        'units' => 'miles'
    ];

    /**
     * Constructor hook method
     *
     * @param array $config The configuration settings provided to this behavior
     * @throws GeoDistanceFatalException If database engine is not MySQL or Postgres
     * @throws GeoDistanceFatalException If latitudeColumn or longitudeColumn are invalid
     * @return void
     */
    public function initialize(array $config)
    {
        $connection = $this->_table->getConnection();

        //ensure database engine is MySQL or Postgres
        if ((!$connection->getDriver() instanceof Mysql) && (!$connection->getDriver() instanceof Postgres)) {
            throw new GeoDistanceFatalException('Only MySQL and Postgres are supported');
        }

        //ensure latitudeColumn and longitudeColumn exist in table
        $collection = $connection->getSchemaCollection();
        $columns = $collection->describe($this->_table->getTable())->columns();
        if (!in_array($this->_config['latitudeColumn'], $columns) ||
            !in_array($this->_config['longitudeColumn'], $columns)) {
            throw new GeoDistanceFatalException('Invalid column');
        }
        $this->_config['latitudeColumn'] = $this->_table->getAlias() . '.' . $this->_config['latitudeColumn'];
        $this->_config['longitudeColumn'] = $this->_table->getAlias() . '.' . $this->_config['longitudeColumn'];
    }

    /**
     * Find By Distance using spherical cosine law
     *
     * @param \Cake\ORM\Query $query Query to modify
     * @param array $options Options for the query
     * @throws GeoDistanceInvalidArgumentException If parameters are missing or invalid
     * @return \Cake\ORM\Query
     */
    public function findByDistance(Query $query, array $options)
    {
        // set up parameters
        $this->_validateOptions($options);

        $latitude = $options['latitude'];
        $longitude = $options['longitude'];
        $radius = $options['radius'];
        if (isset($options['units']) && !empty($options['units'])) {
            $units = $options['units'];
        } else {
            $units = $this->_config['units'];
        }

        $earthRadius = $this->_getEarthRadius($units);

        // construct query
        $sphericalCosineSql = "(:earth_radius * ACOS(
            COS(RADIANS(:latitude)) *
            COS(RADIANS({$this->_config['latitudeColumn']})) *
            COS( RADIANS({$this->_config['longitudeColumn']}) - RADIANS(:longitude) ) +
            SIN(RADIANS(:latitude)) *
            SIN(RADIANS({$this->_config['latitudeColumn']}))
        ) )";

        $connection = $query->getConnection();
        if ($connection->getDriver() instanceof Mysql) {
            $distance = "ROUND($sphericalCosineSql, 3)";
        } elseif ($connection->getDriver() instanceof Postgres) {
            $distance = "ROUND( CAST($sphericalCosineSql AS numeric), 3)";
        }

        $queryOptions = [
            'fields' => ['distance' => $distance],
            'order' => ['distance ASC'],
            'conditions' => ["$distance <= :radius"]
        ];
        $query->find('all', $queryOptions)
        ->bind(':earth_radius', $earthRadius, 'integer')
        ->bind(':latitude', $latitude, 'float')
        ->bind(':longitude', $longitude, 'float')
        ->bind(':radius', $radius, 'float');

        return $query;
    }

    /**
     * Earth Radius
     *
     * The (approximate) mean radius of the earth in miles or km
     * @param string $units miles or km
     * @return float
     */
    private function _getEarthRadius($units)
    {
        if (in_array($units, ['kilometres', 'km'])) {
            return 6371;
        } else {
            return 3958.756;
        }
    }

    /**
     * Validate Options
     *
     * Check the configuration options conform to expected values/formats
     * @param array $options Options for the query
     * @throws GeoDistanceInvalidArgumentException If parameters are missing or invalid
     * @return void
     */
    private function _validateOptions(array $options)
    {
        if (!isset($options['latitude'])) {
            throw new GeoDistanceInvalidArgumentException('Parameter latitude must be specified');
        } else {
            if (($options['latitude'] < -90) || ($options['latitude'] > 90) || !is_numeric($options['latitude'])) {
                throw new GeoDistanceInvalidArgumentException('Parameter latitude must be in the range -90 - 90');
            }
        }

        if (!isset($options['longitude'])) {
            throw new GeoDistanceInvalidArgumentException('Parameter longitude must be specified');
        } else {
            if (($options['longitude'] < -180) || ($options['longitude'] > 180) || !is_numeric($options['longitude'])) {
                throw new GeoDistanceInvalidArgumentException('Parameter longitude must be in the range -180 - 180');
            }
        }

        if (!isset($options['radius'])) {
            throw new GeoDistanceInvalidArgumentException('Parameter radius must be specified');
        } else {
            if (!is_numeric($options['radius'])) {
                throw new GeoDistanceInvalidArgumentException('Parameter radius must be a number');
            }
        }

        if (isset($options['units']) && !empty($options['units'])) {
            if (!in_array($options['units'], ['miles', 'mi', 'kilometres', 'km'])) {
                throw new GeoDistanceInvalidArgumentException(
                    "Parameter units must be one of: 'miles', 'mi', 'kilometres', 'km'"
                );
            }
        }
    }
}
